File: go/anagram/anagram.go
package anagram

import (
	"sort"
	"strings"
)



func Detect(word string, candidates []string) []string {


	//sort all letters of word => use as lookup key
	w := []rune(strings.ToLower(word))
	sort.Slice(w, func(i,j int) bool{ return w[i] < w[j] })


	//go over candidates and compared 'sorted words'
	res := []string{}
	for _, cw := range candidates {

		c := []rune(strings.ToLower(cw))
		sort.Slice(c, func(i,j int) bool { return c[i] < c[j] }) //sort chars

		if string(c) == string(w) && strings.ToLower(word) != strings.ToLower(cw) { //word itself is not an anagram !
			res = append(res, cw)
		}

	}

	return res
}

File: haskell/anagram/src/Anagram.hs
module Anagram (anagramsFor) where

import Data.List (sort)
import Data.Char (toLower)

anagramsFor :: String -> [String] -> [String]
anagramsFor xs xss = filter (\x -> lower xs /= lower x) $ filter (\x ->  hash xs == hash x) xss

hash :: String -> String
hash xs = sort $ lower xs

lower :: String -> String
lower = map toLower